### ISSUE TITLE
Latest master SRU fails to train

### ISSUE THREAD

**Issue by andfoy** [taolei87/sru#54]
Thanks for your awesome work on this model, right now we're using it to process language and visual features to generate segmentation masks from referral expressions, however, we're experiencing major issues with the latest master revision. 

Until commit 43c85ed, our model did train perfectly (without gradient clipping), however, after updating to the latest version, the model fails to converge during training. Also, our old weights are not working as expected and we haven't touched our code at all. We would like to know if is there any incompatibility introduced after the multigpu branch was merged?

Here's a little snippet depicting our current declaration and usage of the SRU model:

```python
class Net(nn.Module):
    def __init__(self, *args, **kwargs):
        ...
        self.lang_model = SRU(emb_size, hid_size, num_layers=lang_layers)
        ...
        self.mrnn = SRU(mixed_size, hid_mixed_size,
                        num_layers=mixed_layers)

    def forward(self, vis, lang):
        ...
        lang = self.emb(lang)
        # LxB representation
        lang = torch.transpose(lang, 0, 1)
        lang, _ = self.lang_model(lang)
        ...
        # input has dimensions: seq_length x batch_size (1) x mix_size
        output, _ = self.mrnn(q)
        ...
```

I would appreciate any guidance to solve this issue. Right now, our model takes 5 days to train and we would like to parallelize it on multiple GPUs, such that we can reduce that time.

**Comment by taolei87** [comment]
Hi,

  - The major change to the mater is commit https://github.com/taolei87/sru/commit/bcc6cde62cdb19f0f4d23a2ca548d8e63fe683c5 , in which a scaling constant term is introduced in the highway transformation:<br>
`h[t] = r[t] * c[t] + (1-r[t]) * x[t] * scale_x`<br>
I've tested this version on classification and SQuaD, and the training is fine.

  - If you have already a trained model, it wouldn't be compatible due to the `scale_x` constant. The work-around is to set `self.scale_x = 1` for each `SRUCell()` instance. Let me know if this works.

  - Does the gradient explode (the training diverges) in the latest master version? I would like to make the new version more stable, so it would be great if you can share more information (e.g. `emb_size`, `hid_size`, `lang_layers` and if the embeddings have large initial values etc.)

  - If you would like to re-train the model and revert the SRU() behavior back to the old version, you can try setting `rescale=False` in the [parameter initialization](https://github.com/taolei87/sru/blob/master/cuda_functional.py#L603) method of SRUCell.

Finally, please note that multi-GPU training speed largely depends on the speed of data transfer between GPUs. Things like [nvlink](https://blogs.nvidia.com/blog/2014/11/14/what-is-nvlink/) are perhaps needed to enable fast model parameter / batch data transfer. You could compare single-GPU vs multi-GPU speed using the script:
https://github.com/taolei87/sru/blob/master/misc/test_multigpu.py

**Comment by andfoy** [comment]
Hi, and thanks for your very thoroughly answer. Here there are some replies and additional questions:

> The major change to the mater is commit bcc6cde , in which a scaling constant term is introduced in the highway transformation

Does this scaling term improves performance on language modelling tasks, because we'd like to use it only on the language SRU, but not in the Multimodal one.

>  If you have already a trained model, it wouldn't be compatible due to the scale_x constant. The work-around is to set self.scale_x = 1 for each SRUCell() instance. Let me know if this works.

I'll take a look at this.

> Does the gradient explode (the training diverges) in the latest master version? I would like to make the new version more stable, so it would be great if you can share more information (e.g. emb_size, hid_size, lang_layers and if the embeddings have large initial values etc.)

Right now we're using 1000 word embedding dimensions, which are randomly initialized and then learnt during the training process. Also, our ``hid_size`` corresponds to 1000 and we use 3 SRU layers. With respect to the gradient explosion, I'll train our model from scratch and I'll tell you the overall behaviour.  

>  If you would like to re-train the model and revert the SRU() behavior back to the old version, you can try setting rescale=False in the parameter initialization method of SRUCell.

Also, I'd like to send a PR to configure this parameter as part of the ``__init__`` constructor of the SRU.

**Comment by taolei87** [comment]
Hi, thank you for the PR!

  - I didn't test the new version on language modeling task. I'm now focusing on translation tasks, and can check on this later. The scaling constant is greater than 1 (when constructed with default arguments), so I guess there's a chance of explosion (although I didn't see it). Perhaps increasing the batch size or tune the learning rate a bit would help.

  - You could check if the embedding values are initialized to be small values, with a variance much smaller than 1.0 (this is the assumption for deriving proper weight initialization such as Xavier / Kaiming init). I used a variance `1/d` in my experiments. 

  - If the inputs to the SRU are normalized using things like BatchNorm or LayerNorm, you can set `SRUCell.layer_norm = True` to [adjust initial weights](https://github.com/taolei87/sru/blob/master/cuda_functional.py#L630-L632) accordingly.

**Comment by andfoy** [comment]
Update: It seems that we got to replicate the old SRU results by using the new one without rescaling parameter

**Comment by taolei87** [comment]
Sounds great!  Let me know if you have more questions / issues. :)
